### ISSUE TITLE
size_of function doesn't work with with type parameters (or else bogus error message about `Z

### ISSUE THREAD

**Issue by mikebenfield** [rust-lang/rust#78223]
I tried this code:

```rust
use std::mem::size_of;
fn f<T: Sized>(t: T) {
    let x: [u8; size_of::<T>()];
}

fn main() {
    f(0u8)
}
```

I expected to see this happen: Either for the code to compile cleanly or to give a sane error message. 

Instead, this happened:
```
error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> trashc.rs:3:27
    |
2   | fn f<T: Sized>(t: T) {
    |      - this type parameter needs to be `Sized`
3   |     let x: [u8; size_of::<T>()];
    |                           ^ doesn't have a size known at compile-time
    |
   ::: /Users/mike/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/mem/mod.rs:303:22
    |
303 | pub const fn size_of<T>() -> usize {
    |                      - required by this bound in `std::mem::size_of`
```

### Meta

`rustc --version --verbose`:
```
rustc 1.49.0-nightly (043eca7f0 2020-10-17)
binary: rustc
commit-hash: 043eca7f0b34d12e61c44206beca740628647080
commit-date: 2020-10-17
host: x86_64-apple-darwin
release: 1.49.0-nightly
LLVM version: 11.0
```

**Comment by lzutao** [comment]
Duplicate of #57144 at least.

**Comment by JohnTitor** [comment]
I'm going to close this as a duplicate. If you have any thoughts on this, feel free to leave a comment on #57144. Thanks for reporting!
